tests/integration/v1/requirer_charm/src/charm.py

```diff
@@ -0,0 +1,187 @@
+    def _replicas_relation_created(self) -> bool:
+        return self._relation_created("replicas")
+
+    def _store_private_key(self, private_key: str) -> None:
+        self._store_item(key="private_key", value=private_key)
+
+    def _store_csr(self, csr: str) -> None:
+        self._store_item(key="csr", value=csr.strip())
+
+    def _store_certificate(self, certificate: str) -> None:
+        self._store_item(key="certificate", value=certificate)
+
+    def _store_ca_certificate(self, certificate: str) -> None:
+        self._store_item(key="ca", value=certificate)
+
+    def _store_ca_chain(self, chain: List[str]) -> None:
+        self._store_item(key="chain", value=json.dumps(chain))
+
+    def _store_item(self, key: str, value: str) -> None:
+        replicas_relation = self.model.get_relation("replicas")
+        if not replicas_relation:
+            raise RuntimeError("Peer relation not created")
+        replicas_relation.data[self.unit].update({key: value})
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns whether given relation was created.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: True/False
+        """
+        try:
+            if self.model.get_relation(relation_name):
+                return True
+            return False
+        except KeyError:
+            return False
+
+    def _get_item_from_peer_relation_data(self, key: str) -> Optional[str]:
+        replicas_relation = self.model.get_relation("replicas")
+        if not replicas_relation:
+            raise RuntimeError("Peer relation not created")
+        return replicas_relation.data[self.unit].get(key, None)
+
+    def _request_certificate(self) -> None:
+        """Requests TLS certificates.
+
+        Returns:
+            None
+        """
+        if not self._stored_private_key:
+            raise RuntimeError("Private key not stored.")
+        csr = generate_csr(
+            private_key=self._stored_private_key.encode(),
+            subject=CONFIG_CHANGED,
+        )
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+        self._store_csr(csr.decode())
+        self.unit.status = MaintenanceStatus("Requesting new certificate")
+
+    def _generate_private_key(self) -> None:
+        """Generates root certificate to be used to sign certificates.
+
+        Returns:
+            None
+        """
+        private_key = generate_private_key()
+        self._store_private_key(private_key.decode())
+        logger.info("Private key generated and stored.")
+
+    def _on_install(self, event: InstallEvent) -> None:
+        if not self._replicas_relation_created:
+            self.unit.status = WaitingStatus("Waiting for replicas relation to be created")
+            event.defer()
+            return
+        self._generate_private_key()
+        self.unit.status = BlockedStatus("Waiting for relation to be created")
```

Comment:
Why aren't we creating the private key in the leader unit and passing it through peer relation? Just like we do in the provider with the root certificates.

Reply:
Because we are only using the private key once (on install)